pkg/image/util/helpers.go

```diff
@@ -14,75 +14,96 @@ import (
+		if len(manifest.History) != len(manifest.FSLayers) {
+			return fmt.Errorf("the image %s (%s) has mismatched history and fslayer cardinality (%d != %d)", image.Name, image.DockerImageReference, len(manifest.History), len(manifest.FSLayers))
+		}
+
+		image.DockerImageLayers = make([]imageapi.ImageLayer, len(manifest.FSLayers))
+		for i, obj := range manifest.History {
+			layer := manifest.FSLayers[i]
+
+			var size imageapi.DockerV1CompatibilityImageSize
+			if err := json.Unmarshal([]byte(obj.DockerV1Compatibility), &size); err != nil {
+				size.Size = 0
+			}
+
+			// reverse order of the layers: in schema1 manifests the
+			// first layer is the youngest (base layers are at the
+			// end), but we want to store layers in the Image resource
+			// in order from the oldest to the youngest.
+			revidx := (len(manifest.History) - 1) - i // n-1, n-2, ..., 1, 0
+
+			image.DockerImageLayers[revidx].Name = layer.DockerBlobSum
+			image.DockerImageLayers[revidx].LayerSize = size.Size
+			image.DockerImageLayers[revidx].MediaType = schema1.MediaTypeManifestLayer
+		}
+	case 2:
+		// The layer list is ordered starting from the base image (opposite order of schema1).
+		// So, we do not need to change the order of layers.
+		image.DockerImageLayers = make([]imageapi.ImageLayer, len(manifest.Layers))
+		for i, layer := range manifest.Layers {
+			image.DockerImageLayers[i].Name = layer.Digest
+			image.DockerImageLayers[i].LayerSize = layer.Size
+			image.DockerImageLayers[i].MediaType = layer.MediaType
+		}
+	default:
+		return fmt.Errorf("unrecognized Docker image manifest schema %d for %q (%s)", manifest.SchemaVersion, image.Name, image.DockerImageReference)
+	}
+
+	if image.Annotations == nil {
+		image.Annotations = map[string]string{}
+	}
+	image.Annotations[imageapi.DockerImageLayersOrderAnnotation] = imageapi.DockerImageLayersOrderAscending
+
+	return nil
+}
+
 // ImageWithMetadata mutates the given image. It parses raw DockerImageManifest data stored in the image and
 // fills its DockerImageMetadata and other fields.
 func ImageWithMetadata(image *imageapi.Image) error {
 	if len(image.DockerImageManifest) == 0 {
 		return nil
 	}
 
+	ReorderImageLayers(image)
+
 	if len(image.DockerImageLayers) > 0 && image.DockerImageMetadata.Size > 0 && len(image.DockerImageManifestMediaType) > 0 {
 		glog.V(5).Infof("Image metadata already filled for %s", image.Name)
-
-		ReorderImageLayers(image)
-
-		// don't update image already filled
 		return nil
 	}
 
-	manifestData := image.DockerImageManifest
-
 	manifest := imageapi.DockerImageManifest{}
-	if err := json.Unmarshal([]byte(manifestData), &manifest); err != nil {
+	if err := json.Unmarshal([]byte(image.DockerImageManifest), &manifest); err != nil {
 		return err
 	}
 
-	if image.Annotations == nil {
-		image.Annotations = map[string]string{}
+	err := fillImageLayers(image, manifest)
+	if err != nil {
+		return err
 	}
 
 	switch manifest.SchemaVersion {
-	case 0:
-		// legacy config object
```

Comment:
on what basis are we confident/safe to remove this?

Reply:
This is for Docker before v1.3.0. Our code relies on DockerImageMetadata and this function for will leave it unfilled schema 0 without an error. We need to fix our code if we still want to have support for schema 0.

Reply:
got it.